contracts/ColonyNetworkENS.sol

```diff
@@ -0,0 +1,112 @@
+/*
+  This file is part of The Colony Network.
+
+  The Colony Network is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  The Colony Network is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with The Colony Network. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+pragma solidity ^0.4.23;
+pragma experimental "v0.5.0";
+
+import "./ens/ENS.sol";
+import "./ColonyNetworkStorage.sol";
+
+
+/// @title First-In-First-Served ENS Registrar
+/// @notice A registrar that allocates subdomains to the first person to claim them.
+/// @notice Source modified from https://github.com/ensdomains/ens/blob/master/contracts/FIFSRegistrar.sol
+contract ColonyNetworkENS is ColonyNetworkStorage {
+
+  bytes32 constant USER_HASH = keccak256("user");
+  bytes32 constant COLONY_HASH = keccak256("colony");
+
+  modifier unowned(bytes32 node, string username) {
+    address currentOwner = ENS(ens).owner(keccak256(abi.encodePacked(node, keccak256(username))));
+    require(currentOwner == 0, "colony-label-already-owned");
+    _;
+  }
+
+  event UserLabelRegistered(address indexed user, bytes32 label);
+  event ColonyLabelRegistered(address indexed colony, bytes32 label);
+
+  function setupRegistrar(address _ens, bytes32 _rootNode) public auth {
+    ens = _ens;
+    rootNode = _rootNode;
+    userNode = keccak256(abi.encodePacked(rootNode, USER_HASH));
+    colonyNode = keccak256(abi.encodePacked(rootNode, COLONY_HASH));
+    ENS(ens).setSubnodeOwner(rootNode, USER_HASH, this);
+    ENS(ens).setSubnodeOwner(rootNode, COLONY_HASH, this);
+  }
+
+  function registerUserLabel(string username, string orbitdb)
+  public
+  unowned(userNode, username)
+  {
+    require(bytes(username).length > 0, "colony-user-label-invalid");
+    require(bytes(userLabels[msg.sender]).length == 0, "colony-user-label-already-owned");
+    bytes32 subnode = keccak256(username);
```

Comment:
The packed encoding of `username` is just `username`, right?

Reply:
yes, but there is a warning thrown atm because `keccak` expects `bytes` only and we're giving it a string. I think this is by design which seems will carry over in 0.5.0 https://github.com/ethereum/solidity/issues/3955

To eliminate the warning explicitly convert via `bytes(userName)`